components/contracts/elusiveTargetArcades.ts

```diff
@@ -0,0 +1,98 @@
+ *     the Free Software Foundation, either version 3 of the License, or
+ *     (at your option) any later version.
+ *
+ *     This program is distributed in the hope that it will be useful,
+ *     but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *     GNU Affero General Public License for more details.
+ *
+ *     You should have received a copy of the GNU Affero General Public License
+ *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+export const orderedETAs = [
+    "ff5b4e53-49ea-4d85-b94e-d3c8b3fc7ab3",
+    "80cf04de-8e0b-4f38-b094-600753e2ac24",
+    "85a67f31-75ce-40f5-a281-7765791f58ca",
+    "a79640cb-331a-41af-abaf-58e629fe0a04",
+    "cab4293f-e359-419d-aa6f-83d91a158cf5",
+    "dd68d30f-3900-415f-bb17-84681a2cd4fc",
+    "07bc9bbf-7cba-4cdf-92bb-3ab57f09b1cc",
+    "223aa1f3-64a1-43c0-b3c8-36aebd7998e4",
+    "9cbdb972-95df-4e0a-be77-7937ec6f2fb0",
+    "afb4ce18-bf9d-443d-85a8-207df9011792",
+    "09ea9b32-dc85-4633-9f6a-c5c4bfded5ac",
+    "479ec396-b96f-4e01-94d9-aedaa0582ed9",
+    "5081776a-86b6-4935-a83e-631a65ba8ee8",
+    "8db8fa33-cb41-4a0d-a2de-e30d884afb95",
+    "be787ec9-e7b9-4984-bb39-fda4c71705ec",
+    "74278853-5990-4058-8972-1f10ad12b6d8",
+    "6516174c-abe0-4cd7-acdb-7cda4c9b5016",
+    "d21e2e91-602c-49d2-9d42-e8bcfb810e9a",
+    "6de90688-ad4c-457e-ae25-c4bbc8f55196",
+    "06f99231-9b1e-483f-8a72-f91efcc2fd2b",
+    "17027c6f-7010-43b8-bb42-1a6846fc0b7b",
+    "09a40d33-4820-454b-89af-a10e0b8d3e08",
+    "7886bed7-3805-46c6-a792-4a0d55758934",
+    "2c2afdfe-d396-451a-a6a3-52aca4ea4f1f",
+    "39cc0603-4348-4dbf-9bd3-733cadf2913c",
+    "b4d555eb-e2c9-40c2-b155-328a7019fd28",
+    "e4b29c19-13b4-471b-b188-cd9c0a788cd0",
+    "90121ee0-0431-4a97-9bc8-8a7e2ca30d65",
+    "7569d5e4-5270-4b69-96c8-e47b99876390",
+    "bfb56fe6-06db-440a-aafe-42eeeb223fa1",
+    "b9f55fc3-c53f-4661-a4b6-9956303422aa",
+    "9e469023-e0c4-42bf-8527-f9fcaf624421",
+]
+
+export const locationsWithETA = [
+    "LOCATION_BANGKOK",
+    "LOCATION_EDGY_FOX",
+    "LOCATION_WET_RAT",
+    "LOCATION_COLORADO",
+    "LOCATION_ANCESTRAL_BULLDOG",
+    "LOCATION_GOLDEN_GECKO",
+    "LOCATION_NEWZEALAND",
+    "LOCATION_HOKKAIDO",
+    "LOCATION_MARRAKECH",
+    "LOCATION_MARRAKECH_NIGHT",
+    "LOCATION_ELEGANT_LLAMA",
+    "LOCATION_PARIS",
+    "LOCATION_COLOMBIA",
+    "LOCATION_COASTALTOWN_MOVIESET",
+    "LOCATION_COASTALTOWN",
+    "LOCATION_NORTHSEA",
+    "LOCATION_NORTHAMERICA",
+]
+
+/**
+ * These are the parent locations with sublocations that have ETAs,
+ * and thus need to show ETA challenges on their challenges page.
+ *
+ * It is possible to determine whether a parent is such a parent location
+ * when filtering out challenges for the parent location, but it is
+ * much more efficient to do it this way.
+ */
+export const parentsWithETA = [
```

Comment:
My fingers are itching by this bit and `locationsWithETA`, seems like premature optimization.

Reply:
I'll defer this to @moonysolari.

Reply:
Both of these arrays are used solely in the `filterChallenge` function. When getting the challenges for a contract or a location, this function is called on every challenge that is either in the parent location or a global challenge.

To do without `locationsWithETA`, we will need to loop over all arcade levels and run `resolveContract` to get their locations every time this array is needed.

To do without `parentsWithETA`, we will need to loop over all locations in `locationsWithETA` and get their parent locations from the config file every time this array is needed.

I wrote them as constants considering the frequency of `filterChallenge` being called, the complexity of acquiring the locations with ETA in real-time, and the likelihood that this list will change. However, this implementation is still up for discussion.

Reply:
But why not build it once at runtime, instead of hardcoding it? This is going to give issues if someone is going to add custom stuff through plugins.

Reply:
Good idea. Addressed as required.

Reply:
You also need to remove the arrays from `elusiveTargetArcades.ts`.

Reply:
LGTM - do we want to get rid of the whole orderedET array as well and just built that runtime too?

Reply:
We can't, there's no ordering data for them.

Reply:
In that case I sign off on this in its current state, good job all :)